hypothesis/lms
# Fix `make services` and `make run-docker`

Slack threads:

* https://hypothes-is.slack.com/archives/C4K6M7P5E/p1642601538068700
* https://hypothes-is.slack.com/archives/C4K6M7P5E/p1644242681546769

# Problem: `make services` is broken 

`make services` (and other commands) fail because the `.devdata.env` file is missing, but you can't run `make devdata` (to create the `.devdata.env` file) without having run `make services` first (because `make devdata` needs the DB) so there's a circular dependency.

You can test this by shutting down your services (`make services args=down`) and deleting your `.devdata.env` file, simulating a new user who hasn't run `make services` or `make devdata` yet. Now try to get the app running:

```
$ make services
ERROR: Couldn't find env file: /home/seanh/Projects/lms/.devdata.env
ERROR: InvocationError for command /home/seanh/Projects/lms/.tox/dockercompose/bin/docker-compose up -d (exited with code 1)
_________________________________________________________________________________________ summary __________________________________________________________________________________________
ERROR:   dockercompose: commands failed
make: *** [Makefile:28: services] Error 1

$ make devdata
...
Traceback (most recent call last):
...
psycopg2.OperationalError: could not connect to server: Connection refused
    Is the server running on host "localhost" (127.0.0.1) and accepting
    TCP/IP connections on port 5433?

(Background on this error at: https://sqlalche.me/e/14/e3q8)
ERROR: InvocationError for command /home/seanh/Projects/lms/.tox/dev/bin/python bin/update_dev_data.py conf/development.ini (exited with code 1)
_________________________________________________________________________________________ summary __________________________________________________________________________________________
ERROR:   dev: commands failed
make: *** [Makefile:46: devdata] Error 1

$ make services
ERROR: Couldn't find env file: /home/seanh/Projects/lms/.devdata.env
ERROR: InvocationError for command /home/seanh/Projects/lms/.tox/dockercompose/bin/docker-compose up -d (exited with code 1)
_________________________________________________________________________________________ summary __________________________________________________________________________________________
ERROR:   dockercompose: commands failed
make: *** [Makefile:28: services] Error 1

$ 
```

# ~Problem: `make run-docker` is broken~

**Edit: This is a result of not running `make docker` first and is un-related.**

The `make run-docker` suffers from the same `make services` and `make devdata` circular dependency as the rest of the commands. Once you resolve that (get your `.devdata.env` file in place and your services running) `make run-docker` is running into another error:

```terminal
$  make run-docker
dockercompose installed: *** listing modules disabled by tox-pip-sync in pyproject.toml ***
dockercompose run-test-pre: PYTHONHASHSEED='1757428560'
dockercompose run-test: commands[0] | docker-compose up --force-recreate web
Pulling web (hypothesis/lms:dev)...
ERROR: manifest for hypothesis/lms:dev not found: manifest unknown: manifest unknown
ERROR: InvocationError for command /home/seanh/Projects/lms/.tox/dockercompose/bin/docker-compose up --force-recreate web (exited with code 1)
___________________________________________________________ summary ____________________________________________________________
ERROR:   dockercompose: commands failed
make: *** [Makefile:108: run-docker] Error 1
```



I actually saw a _different_ error when I tried to run it today but that may have been some local Docker / Docker Compose confusion, or who knows what.

# Solution

This PR's first commit reverts https://github.com/hypothesis/lms/pull/3112 meaning that LMS's `make run-docker` command is now the same as in other projects (and the same as it used to be in LMS). That leaves `make run-docker` with the two problems that https://github.com/hypothesis/lms/pull/3112 meant to address:

## Solution: Docker host neworking

Docker host networking is broken. This PR's second commit fixes that by adding `--add-host host.docker.internal:host-gateway` to the `docker run` command which means that `host.docker.internal` can be used to refer to the host machine in environment variables (e.g. `-e H_API_URL_PRIVATE=http://host.docker.internal:5000/api/`)

## Solution: environment variables

The `docker run` command in the `Makefile` doesn't contain all of the environment variables needed to get the app running. This PR's third commit partially addresses this issue by adding some of the missing envvars to `Makefile` and improving the comment that explains where to look for others.

We can't simply add all the missing envvars to `Makefile` because some of them are "secret" envvars that come from the `.devdata.env` file that `make devdata` creates. A solution might be to get Make to read the environment variables from `.devdata.env` and pass them to Docker or to get Docker to read `.devdata.env` itself.

Even then the problem wouldn't be entirely solved though because not all environment variables come from `.devdata.env`. Some of them come from `tox.ini`. We may want to move all envvars out of `tox.ini` and into `conf/development.ini` (I think settings defined in the `conf/*.ini` files are actually fine: both tox and Docker read these files currently, though I'm not sure whether `make run-docker` reads `conf/production.ini` or `conf/development.ini`).

# Testing

On this branch:

1. Edit the `docker run` command in `Makefile` to add the missing envvars. See https://hypothes-is.slack.com/archives/C4K6M7P5E/p1644242733230589?thread_ts=1644242681.546769&cid=C4K6M7P5E for a minimal set of envvars to add.
2. Run each of the other apps (h, Via, and Via HTML) with the normal `make services`, `make devdata` then `make dev`
3. Run LMS in Docker with `make services`, `make devdata` then `make run-docker`
4. Go to https://hypothesis.instructure.com/ and HTML and PDF assignments should be working. (Things like Google Drive, OneDrive, the admin pages, etc won't be working due to missing envvars.)
